### PR TITLE
Android Pie Keys attestation improvment#958

### PR DESCRIPTION
**Pull request by A-AFTAHI**
Thank you for submitting a Pull Request to the Mobile Security Testing Guide. Please make sure that:

- [ ] Your contribution is written in the 2nd person (e.g. you)
- [ ] Your contribution is written in an active present form for as much as possible.
- [ ] You have made sure that the reference section is up to date (e.g. please add sources you have used, make sure that the references to MITRE/MASVS/etc. are up to date)
- [ ] Your contribution has proper formatted markdown and/or code
- [ ] Any references to website have been formatted as [TEXT](URL “NAME”)
- [ ] You verified/tested the effectiveness of your contribution (e.g.: is the code really an effective remediation? Please verify it works!)

If your PR is related to an issue. Please end your PR test with the following line:
This PR covers issue #958.

### CODE REVIEW

File: Document/0x05e-Testing-Cryptography.md [hunk 1]
```diff
@@ -262,6 +262,24 @@ Now, it is clear that regularly prompting the user for its passphrase is not som
...
+Android Pie adds the ability to import keys securely into the `AndroidKeystore`. First `AndroidKeystore` generates a key pair using `PURPOSE_WRAP_KEY` which should also be protected with an attestation certficate, this pair aims to protect the Keys being imported to `AndroidKeystore`. The encypted keys are generated as ASN.1-encoded message in the `SecureKeyWrapper` format which also contains a description of the ways the imported key is allowed to be used. The keys are then decrypted inside the `AndroidKeystore` hardware belonging to the specific device that generated the wrapping key so they never appear as plaintext in the device's host memory.
+
+![Secure key import into Keystore.](Images/Chapters/0x5e/Android9_secure_key_import_to_keystore.png).
+
+#### decryption only on unlocked devices
+
+For more security Android pie introduces the `unlockedDeviceRequied` flag. By passing `true` to the `setUnlockedDeviceRequired()` method the app prevents its keys stored in `AndroidKeystore` from being decrypted when the device is locked, and it requires the screen to be unlocked before allowing decryption.
+
+#### StrongBox Hardware Security module
+
```
Comment (commjoen) [inline Document/0x05e-Testing-Cryptography.md]: Question: how can i see that strongbox is really used during runtime? can you add that (keyinfo)

File: Document/0x05e-Testing-Cryptography.md [hunk 2]
```diff
@@ -262,6 +262,24 @@ Now, it is clear that regularly prompting the user for its passphrase is not som
...
 
+#### Secure key import into Keystore
+
+Android Pie adds the ability to import keys securely into the `AndroidKeystore`. First `AndroidKeystore` generates a key pair using `PURPOSE_WRAP_KEY` which should also be protected with an attestation certficate, this pair aims to protect the Keys being imported to `AndroidKeystore`. The encypted keys are generated as ASN.1-encoded message in the `SecureKeyWrapper` format which also contains a description of the ways the imported key is allowed to be used. The keys are then decrypted inside the `AndroidKeystore` hardware belonging to the specific device that generated the wrapping key so they never appear as plaintext in the device's host memory.
```
Comment (commjoen) [inline Document/0x05e-Testing-Cryptography.md]: Can we get a code sample of how this works or a link towards the details?

Comment (A-AFTAHI) [reply Document/0x05e-Testing-Cryptography.md]: I add an image explaining the process.
still i will add some code sample.

File: Document/0x05e-Testing-Cryptography.md [hunk 3]
```diff
@@ -262,6 +262,40 @@ Now, it is clear that regularly prompting the user for its passphrase is not som
...
+
+SecureKeyWrapper ::= SEQUENCE {
+    wrapperFormatVersion INTEGER,
+    encryptedTransportKey OCTET_STRING,
+    initializationVector OCTET_STRING,
+    keyDescription KeyDescription,
+    secureKey OCTET_STRING,
+    tag OCTET_STRING
+}
+```
```
Comment (commjoen) [inline Document/0x05e-Testing-Cryptography.md]: can you explain the code a bit more? and perhaps highlight the challenges here for developers that are working with it?

**Comment by commjoen** [comment]
This PR seems to be not related to key attestation, but to key security and importing keys: #950... maybe have a separate PR on key attestation?

File: Document/0x05e-Testing-Cryptography.md [hunk 4]
```diff
@@ -262,6 +262,47 @@ Now, it is clear that regularly prompting the user for its passphrase is not som
...
+
+The code above present the diffrent parameters to be set when generating the encrypted keys in the SecureKeyWrapper format. Check the following [link](https://developer.android.com/reference/android/security/keystore/WrappedKeyEntry) for more details.
```
Comment (commjoen) [inline Document/0x05e-Testing-Cryptography.md]: Can you change the `check the following [link]` to `check the Android documentation on [WrappedKeyEntry]`.

**Comment by commjoen** [comment]
Thanks!

**Pull request merged**